examples/apps/cli/main.c

```diff
@@ -97,3 +98,18 @@ int main(int argc, char *argv[])
 
     return 0;
 }
+
+/*
+ * Provide a "weak" log function the platform code can override if desired.
+ */
+#if !OPENTHREAD_ENABLE_DEBUG_UART_LOG
+OT_TOOL_WEAK
+void otPlatLog(otLogLevel aLogLevel, otLogRegion aLogRegion, const char *aFormat, ...)
+{
+    va_list ap;
+
+    va_start(ap, aFormat);
+    otCliPlatLogv(aLogLevel, aLogRegion, aFormat, ap);
+    va_end(ap);
+}
```

Comment:
If I understand it correctly, with the change in this PR OpenThread code  will always provide an implementation for `otPlatLog()` (it may be `OT_TOOL_WEAKE` definition). As you mentioned the platform code can choose to override this if desired.

While this technically works, I personally think it would be better to force the platform developer to be explicit about it. My concern is that if the platform developer really intends to have her/his own `otPlatLog()` but forgets to implement/override the proper function (or  if the definition uses wrong arguments) the code may compile/build ok and the default internal OT implementation will be picked automatically. 

Here is my proposal:
We add a new config option to enable/disable default log output and also select where it is directed to.


````
/**
 * @def OPENTHREAD_SELECT_DEFAULT_LOG_OUTPUT
 * 
 * Define as 0 to disable any default log outputs. Platform should provide `otLogPlat()` implementation
 * Define as 1 for OpenThread to provide a default implementation for `otPlatLog()` which is
 * tied to either  NCP or CLI stream writes.
 * Define as 2 for OpenThread to use the `otPlatLog()` implementation from DEBUG uart feature 
 * (sa OPENTHREAD_ENABLE_DEBUG_UART)
 * 
 */
 ifndef OPENTHREAD_SELECT_DEFAULT_LOG_OUTPUT
 define OPENTHREAD_SELECT_DEFAULT_LOG_OUTPUT            0
 endif
````
This basically combines the `OPENTHREAD_CONFIG_ENABLE_DEFAULT_LOG_OUTPUT` with the new `OPENTHREAD_ENABLE_DEBUG_UART_LOG`...

Thoughts?